crates/p4-analyzer/src/driver.rs

```diff
@@ -0,0 +1,243 @@
+	Console,
+	#[cfg(test)]
+	Buffer(BufferStruct), // Only want to include this for testing
+}
+
+pub fn console_driver() -> Driver {
+	Driver {
+		in_channel: async_channel::unbounded::<Message>(),
+		out_channel: async_channel::unbounded::<Message>(),
+		source_type: DriverType::Console,
+	}
+}
+
+impl Driver {
+	pub fn new(driver_type: DriverType) -> Driver {
+		match driver_type {
+			DriverType::Console => console_driver(),
+			#[cfg(test)]
+			DriverType::Buffer(buffer) => buffer_driver(buffer),
+		}
+	}
+
+	/// Retrieves a [`MessageChannel`] from which [`Message`] instances can be received from (i.e., `stdin`) and sent to (i.e., `stdout`).
+	pub fn get_message_channel(&self) -> MessageChannel {
+		let (sender, _) = self.out_channel.clone();
+		let (_, receiver) = self.in_channel.clone();
+
+		(sender, receiver)
+	}
+
+	fn sender_task(sender: Sender<Message>, input_source: DriverType) {
+		let match_func = || match input_source.clone() {
+			DriverType::Console => Message::read(&mut stdin().lock()),
+			#[cfg(test)]
+			DriverType::Buffer(mut buffer) => block_on(buffer.message_read()),
+		};
+
+		while let Ok(Some(message)) = match_func() {
+			if sender.send_blocking(message).is_err() {
+				break;
+			}
+		}
+	}
+
+	fn receiver_task(receiver: Receiver<Message>, output_source: DriverType) {
+		let message_send = |message: Message| match output_source.clone() {
+			DriverType::Console => message.write(&mut stdout()),
+			#[cfg(test)]
+			DriverType::Buffer(mut buffer) => block_on(buffer.message_write(message)),
+		};
+
+		while let Ok(message) = receiver.recv_blocking() {
+			if message_send(message).is_err() {
+				break;
+			}
+		}
+	}
+
+	/// Starts executing the [`Driver`] instance.
+	pub async fn start(&self, cancel_token: Arc<CancellationToken>) -> Result<(), OperationCanceled> {
+		let (sender, _) = self.in_channel.clone();
+		let (_, receiver) = self.out_channel.clone();
+
+		let source_type = self.source_type.clone();
+		let source_type2 = self.source_type.clone();
+		let _sender_task = std::thread::spawn(move || Self::sender_task(sender, source_type));
+		let receiver_task = std::thread::spawn(move || Self::receiver_task(receiver, source_type2));
+
+		let (sender, _) = self.in_channel.clone();
+		let (_, receiver) = self.out_channel.clone();
+
+		// Joining to the `receiver_task` will block the current thread. Since this will likely be the main thread it will
+		// prevent the async Futures from being driven forward. `spawn_blocking` allows this blocking code to be taken into
+		// another thread and returns a `Future` that we can then await.
+		task::spawn_blocking(move || {
+			cancel_token.run(
+				|| {
+					sender.close();
+					receiver.close();
+				},
+				|| {
+					receiver_task.join().unwrap();
+
+					match cancel_token.is_canceled() {
+						true => Err(OperationCanceled),
+						_ => Ok(()),
+					}
+				},
+			)
+		})
+		.await
+		.unwrap()
+	}
+}
+
+/// All code below is for DriveType::Buffer(BufferStruct)
```

Comment:
nit: Wondering if this should be moved into a child module of `driver`? I think you can then use the `#[cfg(test]` once at the module level.